### PR TITLE
feat(hi-rag): HRM dual-stream architecture with adaptive pondering (P6 Phase 3)

### PR DESCRIPTION
**Pull request by POWERFULMOVES**
## Summary

Implements **P6 Phase 3** — Enhanced HRM (Hierarchical Reasoning Module) with dual-stream architecture.

### Architecture

```
Fast stream (complexity < 0.3) → Single forward pass, low latency
Deep stream (complexity >= 0.3) → Multi-step reasoning with ACT-style pondering
```

### New Modules

| Module | Lines | Purpose |
|---|---|---|
| `hrm/sidecar.py` | 344 | Dual-stream auto-routing HRMSidecar |
| `hrm/ponder.py` | 236 | Adaptive computation with halt probabilities |
| `hrm/reasoning.py` | 271 | Hierarchical: surface→semantic→structural |
| `tests/test_hrm_dual_stream.py` | 381 | 33 tests (23 pass, 10 torch-dependent skip) |

### Configuration (opt-in)

| Variable | Default | Description |
|---|---|---|
| `HRM_ENABLED` | `false` | Enable enhanced HRM |
| `HRM_FAST_THRESHOLD` | `0.3` | Complexity threshold for fast/deep routing |
| `HRM_MAX_PONDER_STEPS` | `8` | Maximum adaptive computation steps |
| `HRM_HALT_THRESHOLD` | `0.9` | Pondering halt confidence |

### Design
- **Opt-in** — defaults to disabled, zero behavior change
- **Backward compatible** — existing hrm_sidecar.py untouched
- **Torch graceful** — works without torch installed
- **Thread-safe** — RLock for concurrent model access

**Related**: PMOVES-P6, Review Recommendation #6

## Summary by CodeRabbit

* **New Features**
  * Added an opt-in enhanced HRM sidecar providing dual-stream (fast vs deep) processing with adaptive pondering and multi-level reasoning.
  * Exposed runtime configuration flags for enablement, routing threshold, ponder step bounds, and halt threshold.

* **Bug Fixes / Monitoring**
  * Added an admin status endpoint showing legacy and enhanced HRM status and current HRM configuration.

* **Tests**
  * Added comprehensive tests covering sidecar behavior, pondering, reasoning, config defaults, and env-dependent torch handling.

### CODE REVIEW

File: pmoves/services/hi-rag-gateway-v2/app.py [hunk 1]
```diff
@@ -914,6 +914,25 @@ async def _handler(msg):
...
+    _hrm_pkg_dir = str(Path(__file__).resolve().parent / "hrm")
+    if _hrm_pkg_dir not in sys.path:
+        sys.path.insert(0, str(Path(__file__).resolve().parent))
+    from hrm import HRMSidecar, HRMSidecarConfig as _HRMCfg
+    _enhanced_hrm = HRMSidecar(_HRMCfg(
+        enabled=HRM_ENABLED,
+        device=DEVICE,
+        fast_threshold=HRM_FAST_THRESHOLD,
+        max_ponder_steps=HRM_MAX_PONDER_STEPS,
+        halt_threshold=HRM_HALT_THRESHOLD,
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/hi-rag-gateway-v2/app.py]: Fixed in 90846e80. Moved the HRM_* env var definitions (HRM_ENABLED, HRM_FAST_THRESHOLD, HRM_MAX_PONDER_STEPS, HRM_HALT_THRESHOLD) to line 915, immediately above the _enhanced_hrm = HRMSidecar(...) init block. The NameError is no longer possible at import time.

File: pmoves/services/hi-rag-gateway-v2/hrm/ponder.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+                if self._halt_head is not None:
+                    try:
+                        halt_prob = float(self._halt_head(state).mean().item())
+                    except Exception:
+                        halt_prob = 1.0
+
+                # Weight with remainder scheduling (ACT-style)
+                remainder = 1.0 - min(step, self.max_steps) / self.max_steps
+                weight = min(halt_prob, remainder) if step >= self.min_steps else 0.0
+
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/hi-rag-gateway-v2/hrm/ponder.py:176]: Fixed in e71c2b45. The halt condition now compares the raw halt_prob (before remainder capping) against halt_threshold. The weight is still capped by remainder scheduling for state accumulation (ACT-style), but the early-exit decision uses the uncapped halt probability. This allows the halt head to trigger early stopping when confidence is high enough, while remainder scheduling prevents overshooting in the accumulated state.

File: pmoves/services/hi-rag-gateway-v2/hrm/sidecar.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+                    # Pondering: adaptive computation steps
+                    state_tensor, ponder_steps = self.ponder.ponder(
+                        q_tensor, transform=self._model,
+                    )
+
+                    # Hierarchical reasoning
+                    reasoning_result = self.reasoner.reason(state_tensor)
+                    reasoning_trace = reasoning_result.get("trace", [])
+
+                    state = state_tensor.squeeze(0).detach().cpu().tolist()
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/hi-rag-gateway-v2/hrm/sidecar.py]: Fixed in 554ece06. The deep stream now uses reasoning_result["state"] (the reasoner-transformed tensor) as the source for the returned embedding, with a fallback to state_tensor if the key is absent. Hierarchical reasoning now actually affects retrieval quality.

[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/hi-rag-gateway-v2/hrm/sidecar.py]: Fixed in 554ece06. The deep stream now extracts the refined state tensor from reasoning_result["state"] instead of using the original state_tensor. This ensures hierarchical reasoning actually transforms the embedding used for retrieval.

File: pmoves/services/hi-rag-gateway-v2/app.py [hunk 4]
```diff
@@ -914,6 +914,25 @@ async def _handler(msg):
...
+    ))
+    logger.info("Enhanced HRM dual-stream sidecar initialised (enabled=%s)", HRM_ENABLED)
+except Exception:
+    logger.info("Enhanced HRM sidecar not available; using legacy controller only")
+    _enhanced_hrm = None
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/hi-rag-gateway-v2/app.py]: Fixed in 90846e80. Moved the HRM_* env var definitions to line 915, immediately above the enhanced HRM init block. The NameError at import time is no longer possible. The broad except was masking this silently — HRM_ENABLED=true now correctly activates the dual-stream sidecar.

File: pmoves/services/hi-rag-gateway-v2/hrm/ponder.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+                total_state = total_state + weight * state
+                steps_taken = step
+
+                # Halt condition
+                if weight >= self.halt_threshold and step >= self.min_steps:
+                    break
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/hi-rag-gateway-v2/hrm/ponder.py:182]: Fixed in e71c2b45. The halt condition now compares raw halt_prob against halt_threshold for the early-exit decision, keeping the remainder-capped weight only for ACT-style state accumulation. With defaults, early stopping can now fire when the halt head outputs >= 0.9.

File: pmoves/services/hi-rag-gateway-v2/hrm/reasoning.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+            "avg_similarity": round(avg_sim, 4),
+        })
+
+        # Structural level: context spread
+        ctx_spread = self._context_spread(context_embeddings)
+        trace.append({
+            "level": "structural",
+            "action": "spread_analysis",
+            "context_spread": round(ctx_spread, 4),
+        })
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/hi-rag-gateway-v2/hrm/reasoning.py:198]: Fixed in d85c9fdb. reason_from_embeddings() now slices the trace to self.levels before returning: return {'trace': trace[:self.levels]}. A sidecar with levels=1 now correctly reports only the surface-level trace entry in the embedding fallback path.

File: pmoves/services/hi-rag-gateway-v2/hrm/sidecar.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+    def __init__(self, config: Optional[HRMSidecarConfig] = None) -> None:
+        self.config = config or HRMSidecarConfig()
+        self.device = self.config.device
+        self._model: Optional[_StreamEncoder] = None
+        self._complexity_head: Optional[_ComplexityHead] = None
+        self.ponder = PonderingController(
+            min_steps=self.config.min_ponder_steps,
+            max_steps=self.config.max_ponder_steps,
+            halt_threshold=self.config.halt_threshold,
+        )
```
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/hi-rag-gateway-v2/hrm/sidecar.py:138]: Fixed in 554ece06. PonderingController constructor now receives d_model, nhead, dim_feedforward, and dropout from the sidecar config. Non-default model dimensions will propagate correctly to the internal _HaltHead and _StepTransform modules instead of falling back to 128/4/512 defaults.

File: pmoves/services/hi-rag-gateway-v2/hrm/sidecar.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (POWERFULMOVES) [reply pmoves/services/hi-rag-gateway-v2/hrm/sidecar.py]: Fixed in 554ece06. The deep stream now extracts the refined state tensor from reasoning_result['state'] instead of using the original state_tensor. Hierarchical reasoning actually transforms the embedding used for retrieval. Falls back to state_tensor if the reasoner doesn't provide a state key.
